File: script/make/protocol/js/RankWorldProtocol.js
export default class RankWorldProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 19201: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 19202: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 19203: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 19204: {
                return new DataView(view.buffer.slice(0, offset));
            }
            case 19205: {
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 19201: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // object
                    const rank = {type, order, key, value, time, name, serverId};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            case 19202: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 
                    // 等级
                    const level = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 职业
                    const classes = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const other = {level, classes};
                    // object
                    const rank = {type, order, key, value, time, name, serverId, other};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            case 19203: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 
                    // 等级
                    const level = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 职业
                    const classes = view.getUint8(offset, false);
                    offset = offset + 1;
                    // 性别
                    const sex = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const other = {level, classes, sex};
                    // object
                    const rank = {type, order, key, value, time, name, serverId, other};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            case 19204: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 
                    // 等级
                    const level = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 职业
                    const classes = view.getUint8(offset, false);
                    offset = offset + 1;
                    // 性别
                    const sex = view.getUint8(offset, false);
                    offset = offset + 1;
                    // VIP等级
                    const vipLevel = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const other = {level, classes, sex, vipLevel};
                    // object
                    const rank = {type, order, key, value, time, name, serverId, other};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            case 19205: {
                // 排行榜
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Rank
                    // 类型
                    const type = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 排名
                    const order = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 键
                    const key = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 值
                    const value = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 时间
                    const time = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 名字
                    const nameLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const nameArray = new Uint8Array(view.buffer.slice(offset, offset + nameLength));
                    const name = textDecoder.decode(nameArray);
                    offset = offset + nameLength;
                    // 服务器ID
                    const serverId = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 
                    // 等级
                    const level = view.getUint16(offset, false);
                    offset = offset + 2;
                    // 职业
                    const classes = view.getUint8(offset, false);
                    offset = offset + 1;
                    // 性别
                    const sex = view.getUint8(offset, false);
                    offset = offset + 1;
                    // VIP等级
                    const vipLevel = view.getUint8(offset, false);
                    offset = offset + 1;
                    // 头像
                    const avatar = view.getUint8(offset, false);
                    offset = offset + 1;
                    // object
                    const other = {level, classes, sex, vipLevel, avatar};
                    // object
                    const rank = {type, order, key, value, time, name, serverId, other};
                    // add
                    list.push(rank);
                }
                return {list};
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/cs/RankWorldProtocol.cs
public static class RankWorldProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Collections.Generic.Dictionary<System.String, System.Object> data) 
    {
        switch (protocol) 
        {
            case 19201:
            {
                return;
            }
            case 19202:
            {
                return;
            }
            case 19203:
            {
                return;
            }
            case 19204:
            {
                return;
            }
            case 19205:
            {
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Collections.Generic.Dictionary<System.String, System.Object> Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 19201:
            {
                // 排行榜
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // Rank
                    // 类型
                    var type = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 排名
                    var order = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 键
                    var key = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 值
                    var value = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 时间
                    var time = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 名字
                    var nameLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var name = encoding.GetString(reader.ReadBytes(nameLength));
                    // 服务器ID
                    var serverId = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // object
                    var rank = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"type", type}, {"order", order}, {"key", key}, {"value", value}, {"time", time}, {"name", name}, {"serverId", serverId}};
                    // add
                    list.Add(rank);
                }
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"list", list}};
            }
            case 19202:
            {
                // 排行榜
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // Rank
                    // 类型
                    var type = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 排名
                    var order = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 键
                    var key = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 值
                    var value = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 时间
                    var time = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 名字
                    var nameLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var name = encoding.GetString(reader.ReadBytes(nameLength));
                    // 服务器ID
                    var serverId = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 
                    // 等级
                    var level = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 职业
                    var classes = reader.ReadByte();
                    // object
                    var other = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"level", level}, {"classes", classes}};
                    // object
                    var rank = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"type", type}, {"order", order}, {"key", key}, {"value", value}, {"time", time}, {"name", name}, {"serverId", serverId}, {"other", other}};
                    // add
                    list.Add(rank);
                }
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"list", list}};
            }
            case 19203:
            {
                // 排行榜
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // Rank
                    // 类型
                    var type = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 排名
                    var order = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 键
                    var key = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 值
                    var value = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 时间
                    var time = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 名字
                    var nameLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var name = encoding.GetString(reader.ReadBytes(nameLength));
                    // 服务器ID
                    var serverId = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 
                    // 等级
                    var level = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 职业
                    var classes = reader.ReadByte();
                    // 性别
                    var sex = reader.ReadByte();
                    // object
                    var other = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"level", level}, {"classes", classes}, {"sex", sex}};
                    // object
                    var rank = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"type", type}, {"order", order}, {"key", key}, {"value", value}, {"time", time}, {"name", name}, {"serverId", serverId}, {"other", other}};
                    // add
                    list.Add(rank);
                }
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"list", list}};
            }
            case 19204:
            {
                // 排行榜
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // Rank
                    // 类型
                    var type = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 排名
                    var order = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 键
                    var key = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 值
                    var value = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 时间
                    var time = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 名字
                    var nameLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var name = encoding.GetString(reader.ReadBytes(nameLength));
                    // 服务器ID
                    var serverId = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 
                    // 等级
                    var level = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 职业
                    var classes = reader.ReadByte();
                    // 性别
                    var sex = reader.ReadByte();
                    // VIP等级
                    var vipLevel = reader.ReadByte();
                    // object
                    var other = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"level", level}, {"classes", classes}, {"sex", sex}, {"vipLevel", vipLevel}};
                    // object
                    var rank = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"type", type}, {"order", order}, {"key", key}, {"value", value}, {"time", time}, {"name", name}, {"serverId", serverId}, {"other", other}};
                    // add
                    list.Add(rank);
                }
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"list", list}};
            }
            case 19205:
            {
                // 排行榜
                var listLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var list = new System.Collections.Generic.List<System.Object>(listLength);
                while (listLength-- > 0)
                {
                    // Rank
                    // 类型
                    var type = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 排名
                    var order = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 键
                    var key = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 值
                    var value = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 时间
                    var time = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 名字
                    var nameLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var name = encoding.GetString(reader.ReadBytes(nameLength));
                    // 服务器ID
                    var serverId = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 
                    // 等级
                    var level = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // 职业
                    var classes = reader.ReadByte();
                    // 性别
                    var sex = reader.ReadByte();
                    // VIP等级
                    var vipLevel = reader.ReadByte();
                    // 头像
                    var avatar = reader.ReadByte();
                    // object
                    var other = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"level", level}, {"classes", classes}, {"sex", sex}, {"vipLevel", vipLevel}, {"avatar", avatar}};
                    // object
                    var rank = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"type", type}, {"order", order}, {"key", key}, {"value", value}, {"time", time}, {"name", name}, {"serverId", serverId}, {"other", other}};
                    // add
                    list.Add(rank);
                }
                return new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"list", list}};
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}